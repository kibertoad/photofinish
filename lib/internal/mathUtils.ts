import { mean } from 'stats-lite'

const percentile2 = require('percentile')

export function filterSignificantSamples(samples: number[]): number[] {
  // return percentile(samples, 5)
  const percentileTopCutoff = percentile2(85, samples)
  const percentileLowCutoff = percentile2(15, samples)

  return samples.filter((entry) => {
    return entry >= percentileLowCutoff && entry <= percentileTopCutoff
  })
}

export function calculateMean(samples: number[]): number {
  return mean(samples)
}
