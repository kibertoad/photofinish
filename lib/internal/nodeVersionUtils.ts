export const nodeVersion = process.versions.node
export const nodeVersionMajor = nodeVersion.slice(0, 2)
export const runtimeVersion = `${nodeVersion}, V8 ${process.versions.v8}`
