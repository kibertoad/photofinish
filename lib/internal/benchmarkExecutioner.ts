import { AsyncFunctionType, Benchmark, FunctionType } from '../BenchmarkBuilder'
import { Measurement } from '../Measurement'
import { filterSignificantSamples, calculateMean } from './mathUtils'

export type BenchmarkResults = {
  benchmarkName: string
  benchmarkEntryName: string
  meanTime: Measurement
}

async function warmupAsync(warmupCycles: number, functionUnderTest: AsyncFunctionType) {
  for (let i = 0; i < warmupCycles; i++) {
    await functionUnderTest()
  }
}

function warmup(warmupCycles: number, functionUnderTest: FunctionType) {
  for (let i = 0; i < warmupCycles; i++) {
    functionUnderTest()
  }
}

export function executeBenchmarkSync(benchmark: Benchmark): BenchmarkResults {
  if (!benchmark.functionUnderTest) {
    throw new Error('Function under test is not set')
  }
  warmup(benchmark.warmupCycles, benchmark.functionUnderTest)

  // Perform all cycles
  const significantSamples = []
  for (let i = 0; i < benchmark.benchmarkCycles; i++) {
    // Perform full cycle
    const samples = []
    for (let i = 0; i < benchmark.benchmarkCycleSamples; i++) {
      const hrStart = process.hrtime()
      benchmark.functionUnderTest()
      const hrEnd = process.hrtime(hrStart)
      const [, timeTakenInNanoSeconds] = hrEnd
      samples.push(timeTakenInNanoSeconds)
    }
    const significantCycleSamples = filterSignificantSamples(samples)
    significantSamples.push(...significantCycleSamples)
  }

  const meanTime = calculateMean(significantSamples)

  return {
    benchmarkName: benchmark.benchmarkName,
    benchmarkEntryName: benchmark.benchmarkEntryName,
    meanTime: new Measurement(meanTime),
  }
}

export async function executeBenchmarkAsync(benchmark: Benchmark): Promise<BenchmarkResults> {
  if (!benchmark.asyncFunctionUnderTest) {
    throw new Error('Async function under test is not set')
  }
  await warmupAsync(benchmark.warmupCycles, benchmark.asyncFunctionUnderTest)

  // Perform all cycles
  const significantSamples = []
  for (let i = 0; i < benchmark.benchmarkCycles; i++) {
    // Perform full cycle
    const samples = []
    for (let i = 0; i < benchmark.benchmarkCycleSamples; i++) {
      const hrStart = process.hrtime()
      await benchmark.asyncFunctionUnderTest()
      const hrEnd = process.hrtime(hrStart)
      const [, timeTakenInNanoSeconds] = hrEnd
      samples.push(timeTakenInNanoSeconds)
    }
    const significantCycleSamples = filterSignificantSamples(samples)
    significantSamples.push(...significantCycleSamples)
  }

  const meanTime = calculateMean(significantSamples)

  return {
    benchmarkName: benchmark.benchmarkName,
    benchmarkEntryName: benchmark.benchmarkEntryName,
    meanTime: new Measurement(meanTime),
  }
}
