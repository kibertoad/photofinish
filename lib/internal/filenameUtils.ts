import { nodeVersionMajor } from './nodeVersionUtils'
import { BenchmarkResults } from './benchmarkExecutioner'

export function generateResultsFilename(benchmarkResults: BenchmarkResults) {
  const labelSuffix = benchmarkResults.label ? `-${benchmarkResults.label}` : ''

  return `${normalizePath(benchmarkResults.benchmarkName)}-${normalizePath(
    benchmarkResults.benchmarkEntryName
  )}-Node_${nodeVersionMajor}${labelSuffix}.json`
}

function normalizePath(path: string) {
  return path.replace(/ /g, '_')
}
