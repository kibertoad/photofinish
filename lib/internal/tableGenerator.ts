const table = require('markdown-table')

import { Benchmark } from '../BenchmarkBuilder'
import { BenchmarkResults } from './benchmarkExecutioner'

export function generateTable(
  benchmark: Benchmark,
  results: readonly BenchmarkResults[]
): string[][] {
  const tableRows: string[][] = [['Option', 'Msecs/op', 'Ops/sec']]
  results.map((entry) => {
    tableRows.push([
      entry.benchmarkEntryName,
      entry.meanTime.getTextInMsecs(),
      entry.meanTime.getTextOpsPerSec(),
    ])
  })

  return table(tableRows)
}
