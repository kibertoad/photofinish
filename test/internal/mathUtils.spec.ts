import { calculateMean, filterSignificantSamples } from '../../lib/internal/mathUtils'

describe('mathUtils', () => {
  describe('getSignificantSamples', () => {
    it('calculates significant samples correctly', () => {
      const samples = [1, 4, 5, 6, 4, 4, 4, 4, 4, 5, 6, 5, 6, 10, 1111, 2222]

      const significantSamples = filterSignificantSamples(samples)

      expect(significantSamples).toMatchSnapshot()
    })
  })

  describe('getSampleMean', () => {
    it('calculates mean correctly', () => {
      const samples = [1, 4, 5, 6, 4, 4, 4, 4, 4, 10, 1111, 99999]

      const mean = calculateMean(samples)

      expect(mean).toMatchSnapshot()
    })
  })
})
