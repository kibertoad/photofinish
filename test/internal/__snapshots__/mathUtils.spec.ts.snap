// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`mathUtils getSampleMean calculates mean correctly 1`] = `8429.666666666666`;

exports[`mathUtils getSignificantSamples calculates significant samples correctly 1`] = `
Array [
  4,
  5,
  6,
  4,
  4,
  4,
  4,
  4,
  5,
  6,
  5,
  6,
  10,
]
`;
